File: src/main/java/net/atos/zac/app/detacheddocuments/converter/RestDetachedDocumentConverter.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.detacheddocuments.converter;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import jakarta.inject.Inject;

import net.atos.zac.app.detacheddocuments.model.RestDetachedDocument;
import net.atos.zac.document.detacheddocument.model.DetachedDocument;
import nl.info.zac.app.identity.converter.RestUserConverter;
import nl.info.zac.enkelvoudiginformatieobject.EnkelvoudigInformatieObjectLockService;
import nl.info.zac.enkelvoudiginformatieobject.model.EnkelvoudigInformatieObjectLock;

public class RestDetachedDocumentConverter {

    private RestUserConverter userConverter;
    private EnkelvoudigInformatieObjectLockService lockService;

    @SuppressWarnings("unused")
    public RestDetachedDocumentConverter() {
        // Default constructor for CDI
    }

    @Inject
    public RestDetachedDocumentConverter(
            final RestUserConverter userConverter,
            final EnkelvoudigInformatieObjectLockService lockService
    ) {
        this.userConverter = userConverter;
        this.lockService = lockService;
    }

    public RestDetachedDocument convert(final DetachedDocument document, final UUID informatieobjectTypeUUID) {
        final RestDetachedDocument restDocument = new RestDetachedDocument();
        restDocument.id = document.getId();
        restDocument.documentUUID = document.documentUUID;
        restDocument.documentID = document.documentID;
        restDocument.informatieobjectTypeUUID = informatieobjectTypeUUID;
        restDocument.titel = document.titel;
        restDocument.zaakID = document.zaakID;
        restDocument.creatiedatum = document.creatiedatum;
        restDocument.bestandsnaam = document.getBestandsnaam();
        restDocument.ontkoppeldDoor = userConverter.convertUserId(document.ontkoppeldDoor);
        restDocument.ontkoppeldOp = document.ontkoppeldOp;
        restDocument.reden = document.getReden();
        final EnkelvoudigInformatieObjectLock lock = lockService.findLock(document.documentUUID);
        restDocument.isVergrendeld = lock != null && lock.getLock() != null;
        return restDocument;
    }

    public List<RestDetachedDocument> convert(
            final List<DetachedDocument> documenten,
            final List<UUID> informatieobjectTypeUUIDs
    ) {
        List<RestDetachedDocument> list = new ArrayList<>();
        for (int index = 0; index < documenten.size(); index++) {
            list.add(convert(
                    documenten.get(index),
                    informatieobjectTypeUUIDs.get(index)
            ));
        }
        return list;
    }
}

File: src/main/kotlin/nl/info/zac/app/detacheddocuments/converter/RestDetachedDocumentConverter.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.app.detacheddocuments.converter

import jakarta.inject.Inject
import nl.info.zac.app.detacheddocuments.model.RestDetachedDocument
import nl.info.zac.app.identity.converter.RestUserConverter
import nl.info.zac.document.detacheddocument.repository.model.DetachedDocument
import nl.info.zac.enkelvoudiginformatieobject.EnkelvoudigInformatieObjectLockService
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor
import java.util.UUID

@AllOpen
@NoArgConstructor
class RestDetachedDocumentConverter @Inject constructor(
    private val userConverter: RestUserConverter,
    private val lockService: EnkelvoudigInformatieObjectLockService
) {
    fun convert(detachedDocument: DetachedDocument, informatieobjectTypeUUID: UUID): RestDetachedDocument {
        val lock = lockService.findLock(detachedDocument.documentUUID)
        return RestDetachedDocument(
            // conversion is always done from an existing detached document (in the database),
            // so id is always present
            id = detachedDocument.id!!,
            documentUUID = detachedDocument.documentUUID,
            documentID = detachedDocument.documentID,
            informatieobjectTypeUUID = informatieobjectTypeUUID,
            titel = detachedDocument.titel,
            zaakID = detachedDocument.zaakID,
            creatiedatum = detachedDocument.creatiedatum,
            bestandsnaam = detachedDocument.bestandsnaam,
            ontkoppeldDoor = userConverter.convertUserId(detachedDocument.ontkoppeldDoor),
            ontkoppeldOp = detachedDocument.ontkoppeldOp,
            reden = detachedDocument.reden,
            isVergrendeld = lock != null && lock.lock != null
        )
    }

    fun convert(documenten: List<DetachedDocument>, informatieobjectTypeUUIDs: List<UUID>): List<RestDetachedDocument> =
        documenten.indices.map { index -> convert(documenten[index], informatieobjectTypeUUIDs[index]) }
}
